app/controllers/ajax_controller/primers.rb

```diff
@@ -0,0 +1,35 @@
+# frozen_string_literal: true
+
+# see ajax_controller.rb
+class AjaxController
+  # Get list of names for autocompletion in mobile app.
+  def name_primer
+    render(json: name_list)
+  end
+
+  # Get list of locations for autocompletion in mobile app.
+  def location_primer
+    render(json: location_list)
+  end
+
+  private
+
+  def name_list
+    name_ids = Observation.select(:name_id).distinct
+    fields = [:id, :text_name, :author, :deprecated, :synonym_id]
+    names = Name.where(id: name_ids).select(*fields)
+    synonyms = names.to_a.map(&:synonym_id).reject(&:nil?).uniq
+    names |= Name.where(deprecated: false, synonym_id: synonyms).select(*fields)
+    names.map do |name|
```

Comment:
I may not have context for why we're returning arrays instead of objects, but if we omit lines 23-26 the app will have the field names returned with the values, which would be helpful for the mobile app.

`[{"id":139785830,"text_name":"Chlorophyllum rachodes","deprecated":false,"synonym_id":402632615,"author":"(Vittad.) Vellinga"},...]`

I've tried the same for the location list (omitting the `map`) and get a `null` id which is expected for what it looks like that query is doing, but `location_id` and `where` come through fine.

`[{"id":null,"location_id":547147019,"where":"Burbank, California, USA"},...]`

Reply:
We can definitely do that... but two thoughts, and these are just personal opinions:

1. I always cringe to expose internals to the outside like that.  Granted, in practice, it is extremely unlikely that the column names in our tables would ever change.  But it's not impossible.  For example, rails made some change that required that we change the content of the rank column of names, and that would have required the changing of everything downstream.  I could imagine a different version of mysql or if we were to change to postgresql then maybe "true" and "false" might change to "0" and "1" or maybe "null" or who knows.  I realize this mapping requires extra code and processing power, but it allows us to guarantee consumers that this is what the result will look like, period, and will never change.

2. I'm also concerned about size, believe it or not.  We're talking 100s of thousands of names.  Adding ... 52 characters? per line, almost doubling the download size, is not something to be dismissed out of hand.  Of course, if we had nginx configured right (maybe we do?) this is compressed before transmission, and the difference in bandwidth would probably be negligible in practice...

Just to throw it out there, sort of a solution that might make everyone happy: Can I just explicitly encode the output as an array of hashes instead?

    ...map do |obs|
      {id: obs.location_id, name: obs.where}
    end

I think this would age better.  I agree that the map for names would be entirely superfluous at the moment. If our internals change, we can always add a map *then* that would preserve the old key names, etc.

(The reason we aren't dealing with objects is that the instantiation of that many Name and Location objects in ActiveRecord makes this query take way longer.  I actually got impatient and never let the query finish when I tried it.  It may even time out for all I know.  Using selects like this is pretty efficient, at least last I tried it.)

Reply:
Thanks for the context! I think a hash is what I meant, in the JSON sense, rather than the Ruby objects? I can work with whatever is available, ultimately. I hadn't considered the size of adding names for values once we were dealing with the production level of names in the database, and that's a fair concern.